### ISSUE TITLE
Newstylization of twisted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4324]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)| @fijal reported|
|-|-|
|Trac ID|trac#4324|
|Type|enhancement|
|Created|2010-02-27 01:33:28Z|
|Branch|https://github.com/twisted/twisted/tree/newstyle-port-4324|

Twisted in places has old style classes. In places they're simply old style classes, in other places they have been hacked over, having a mixture of old and new style in mro. This is particularly unfriendly for PyPy (old style classes being unfriendly, but less so) and prevents Py3k conversion (a bit less of my concern).

Here are some benchmarks on names.py from https://code.launchpad.net/~exarkun/+junk/twisted-benchmarks (run at revision 14). For reference, on this machine CPython scores almost equally 11000 req/sec. Tests were run in chunks of 5 for 5s each (names.py -n 5 -d 5).

PyPy with JIT:

```
7978 names/sec (39890 names in 5 seconds)
13720 names/sec (68600 names in 5 seconds)
14412 names/sec (72060 names in 5 seconds)
14178 names/sec (70890 names in 5 seconds)
14616 names/sec (73080 names in 5 seconds)
```

twisted.internet.protocol.AbstractDatagramProtocol being new-style class

```
8684 names/sec (43420 names in 5 seconds)
14704 names/sec (73520 names in 5 seconds)
14978 names/sec (74890 names in 5 seconds)
15000 names/sec (75000 names in 5 seconds)
15122 names/sec (75610 names in 5 seconds)
```

the same as above plus all base classes of twisted.internet.udp.Port becoming newstyle classes (twisted.persisted.styles.Ephemeral and twisted.python.log.Logger).

```
9250 names/sec (46250 names in 5 seconds)
15840 names/sec (79200 names in 5 seconds)
16328 names/sec (81640 names in 5 seconds)
16544 names/sec (82720 names in 5 seconds)
16302 names/sec (81510 names in 5 seconds)
```

The latter change is probably slightly more controversial, since some of those classes are exported, but I don't know how crucial that is.

<details><summary>Searchable metadata</summary>

```
trac-id__4324 4324
type__enhancement enhancement
reporter__fijal fijal
priority__normal normal
milestone__ 
branch__branches_newstyle_port_4324 branches/newstyle-port-4324
branch_author__fijal fijal
status__closed closed
resolution__fixed fixed
component__core core
keywords__pypy pypy
time__1267234408000000 1267234408000000
changetime__1282495059000000 1282495059000000
version__None None
owner__fijal fijal

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Making `twisted.internet.udp.Port` new-style sounds alright to me.  It's ostensibly public, but it's not something anyone is encouraged to use directly (and there's little or no guarantee that doing so will continue to work, anyway).

`Ephemeral` and `Logger` are a bit trickier.  Application developers are encouraged to subclass these.  Fortunately, neither provides much functionality that's of actual utility.  Thus, I would suggest removing `Logger` from the list of base classes and instead just implementing `logPrefix` (on `FileDescriptor`, I suppose?).  Likewise, `Ephemeral` is largely a debugging aid and I'm not sure anyone would be significantly opposed to just removing it from the list of base classes.  I think those two changes would be sufficient to make `Port` fully new-style.

`AbstractDatagramProtocol` is another thing that's offered for application developers to subclass.  We probably should avoid making it new-style.

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_2"></a>|@fijal set owner to @fijal|
|-|-|

That sounds fine to me. Also as described above, Port is more important than DatagramProtocol (which is purely old style).

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_3"></a>|@fijal commented|
|-|-|

(In [28562]) Branching to 'newstyle-port-4324'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_1):
> Making `twisted.internet.udp.Port` new-style sounds alright to me.  It's ostensibly public, but it's not something anyone is encouraged to use directly (and there's little or no guarantee that doing so will continue to work, anyway).

As one of the few people who is actually subclassing `Port`, I think can endorse this.  I very much doubt it's going to break anything for us.

> `Ephemeral` and `Logger` are a bit trickier.  Application developers are encouraged to subclass these.  Fortunately, neither provides much functionality that's of actual utility.  Thus, I would suggest removing `Logger` from the list of base classes and instead just implementing `logPrefix` (on `FileDescriptor`, I suppose?).  Likewise, `Ephemeral` is largely a debugging aid and I'm not sure anyone would be significantly opposed to just removing it from the list of base classes.  I think those two changes would be sufficient to make `Port` fully new-style.
>
> `AbstractDatagramProtocol` is another thing that's offered for application developers to subclass.  We probably should avoid making it new-style.

It would be really nice to at least allow application developers to start making this transition, and provide public, fully new-style versions of all of these classes.  Given that all the code is going to be the same, perhaps we could move the methods to the new-style version, then have a classic class created by copying the `__dict__` or something?

I don't think that cramming all this work into one giant ticket is good, and the summary is too broad here.  But I'm not sure how I would want to break it up.

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)|@fijal removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix set owner to @fijal|
|-|-|

The only problem I see is that the news file talks about "various classes" but only FileDescriptor is changed. Other than that it looks fine, going by exarkun's arguments above.

Can you make the news file say:

"abstract.FileDescriptor is now a purely new-style object, which helps performance on PyPy"

and then merge

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)|@fijal set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Fixed in r28574:

Merge newstyle-port-4324

Author: fijal
Reviewer: radix
Fixes: [#4324](https://github.com/twisted/twisted/issues/4324)

Remove the classic classes from the bases of FileDescriptor, making it
purely new-style.  This is a performance improvement on PyPy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29937]) Make the same change to twisted/internet/iocpreactor/abstract.py as was made to twisted/internet/abstract.py in r28574

refs [#4617](https://github.com/twisted/twisted/issues/4617)
refs [#4324](https://github.com/twisted/twisted/issues/4324)
